cmake: replace propagated ps input with abspath patch

rhelmot opened the pull request:
## Description of changes

part of https://github.com/NixOS/nixpkgs/pull/296581

Similar to the Darwin case. Should be self-explanatory.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-freebsd
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rhelmot committed:
cmake: propagate ps as a build input on native FreeBSD

AndersonTorres commented on pkgs/by-name/cm/cmake/package.nix:
```diff
@@ -101,7 +101,7 @@ stdenv.mkDerivation (finalAttrs: {
   ++ lib.optional (stdenv.isDarwin && !isMinimalBuild) SystemConfiguration;
 
-  propagatedBuildInputs = lib.optional stdenv.isDarwin ps;
+  propagatedBuildInputs = lib.optional (stdenv.isDarwin || stdenv.isFreeBSD) ps;
```
Why not propagate it over Linux too?

alyssais replied:
Propagation is weird and difficult to debug and should be avoided where possible.

What's this used for?  The library or the executable?  If executable, could we wrap instead?

rhelmot replied:
Looks to be both. However, the place it's used is extremely well isolated and unlikely to change (it's a vendored lib), so we can assuredly just patch it to use an absolute path. This can remove the propagation on Darwin as well.

I suspect ps is part of stdenv on linux so it is not needed there, so I will make the patch only apply to Darwin | FreeBSD in order to minimize rebuilds.

rhelmot committed:
cmake: remove propagatedBuildInput ps and patch to use an abspath

This would be an approprate patch on all platforms, but it seems to
trigger infinite recursion.

alyssais commented on pkgs/by-name/cm/cmake/package.nix:
```diff
@@ -66,7 +67,13 @@ stdenv.mkDerivation (finalAttrs: {
+    substituteAll {
+      src = ./007-darwin-bsd-ps-abspath.diff;
+      ps = "${lib.getBin ps}/bin/ps";
+    });
```
This might be more concisely done with a substituteInPlace or sed.

rhelmot replied:
I was under the impression that this was the kosher way to do this, since direct comparisons are hard to debug if the targeted code changes.

alyssais replied:
Since the patch is conditional, doing it this way means that it's quite likely that Linux users will update the package, without realising that the patch no longer applies, meaning the package ends up not even building on FreeBSD and Darwin.

(OfBorg won't help here, because CMake updates have to go to staging, and OfBorg is generally useless on staging because it can never get past building LLVM without timing out.)

rhelmot replied:
isn't not building preferable to failing in mysterious ways down the line?

rhelmot replied:
I can also make it unconditional on linux, but this will be another mass rebuild.

AndersonTorres replied:
A mass rebuild will happen anyway in other machines.

alyssais replied:
> isn't not building preferable to failing in mysterious ways down the line?

That can happen either way, if they start using it elsewhere in the file, for example.

> I can also make it unconditional on linux, but this will be another mass rebuild.

Probably best not to if it would bring in a dependency that doesn't get used.

alyssais replied:
But I suppose even with sed or substituteInPlace we'd have the same risk of it not being caught due to having to do it conditionally, so I guess this is fine.

rhelmot commented:
I went ahead and made the patch unconditional on all platforms. ps won't be unused on linux, it'll just override the stdenv ps.

rhelmot commented:
On never mind. It looks like that triggers infinite recursion on linux.

SuperSandro2000 commented on pkgs/by-name/cm/cmake/package.nix:
```diff
@@ -66,7 +67,13 @@ stdenv.mkDerivation (finalAttrs: {
+  ++ lib.optional (stdenv.isDarwin || stdenv.isFreeBSD) (
+    substituteAll {
+      src = ./007-darwin-bsd-ps-abspath.diff;
+      ps = "${lib.getBin ps}/bin/ps";
```
```suggestion
      ps = lib.getExe ps;
```

rhelmot replied:
looks good, committed.

alyssais reviewed (approved):
Looks good!  The change to propagation means it's possibly a breaking change, which is currently only allowed in staging, so I'll merge this there.
